dai_cadcad/keeper.py

```diff
@@ -0,0 +1,101 @@
+""" Keeper Module
+    Class-based representation of a keeper in the Maker ecosystem.
+    By this, we mean any external agent, not just auction keepers.
+    (contains only what is necessary for the simulation)
+"""
+
+from uuid import uuid4
+
+from dai_cadcad.pymaker.numeric import Wad  # , Rad, Ray
+
+# from dai_cadcad.util import require
+
+
+class Keeper:
+    ADDRESS = ""
+
+    vat = None
+    dai_join = None
+    gem_joins = {}
+    flippers = {}
+
+    urns = {}
+    num_urns = 0
+
+    def __init__(self, vat, dai_join, ilks):
+
+        self.ADDRESS = f"keeper-{uuid4().hex}"
+
+        self.vat = vat
+        self.dai_join = dai_join
+
+        for ilk in ilks:
+            self.gem_joins[ilk["ilk_id"]] = ilk["gem_join"]
+            self.flippers[ilk["ilk_id"]] = ilk["flipper"]
+
+    def generate_urn_key(self):
+        urn_key = f"urn-{self.num_urns}-{self.ADDRESS}"
+        self.num_urns += 1
+        return urn_key
+
+    def open_vault(self, ilk_id, dink, dart):
+        urn_key = self.generate_urn_key()
+        self.urns[urn_key] = ilk_id
+        # Lock collateral into the system
+        self.gem_joins[ilk_id].join(self.ADDRESS, self.ADDRESS, dink)
+        # Create a vault with the locked collateral
+        # `vat.gem` and `vat.dai` use the Keeper's address, while `vat.urns` uses `urn_key`
+        self.vat.frob(ilk_id, urn_key, self.ADDRESS, self.ADDRESS, dink, dart)
+        # TODO: Ensure that `vat.move` can create a recipient address if it doesn't exist
+        # Move in-system DAI from the urn address to the keeper address
+        self.vat.move(urn_key, self.ADDRESS)
+
+    def close_vault(self, urn_key):
+        ilk_id = self.urns[urn_key]
+        dink = Wad(0) - self.vat.urns[ilk_id][urn_key]["ink"]
+        dart = Wad(0) - self.vat.urns[ilk_id][urn_key]["art"]
+        # Zero out the vault, freeing collateral
+        self.vat.frob(ilk_id, urn_key, self.ADDRESS, self.ADDRESS, dink, dart)
+        # Withdraw freed collateral
+        self.gem_joins[ilk_id].exit(self.ADDRESS, self.ADDRESS, Wad(0) - dink)
+        del self.urns[urn_key]
+
+    # def keeper_bid_flipper_eth(
+    #     self, flipper, vat, spotter, stance, bid_id, user_id, now, stats
+    # ):
+    #     # TODO: Reference flipper_dent, flipper_tend, maybe as instance variables or import
```

Comment:
The relevant Flipper will be an object saved to `self.flippers[ilk_id]` (or, if you have other suggestions for keepers keeping track of flippers, i'm open to em), so we'll call Flipper functions using something like `self.flippers[ilk_id].dent(...)`

Reply:
(I'll omit other comments like "we should use `self.flippers[ilk_id]` here" in the review since it's already been mentioned)